packages/server/src/fhir/lookups/reference.ts

```diff
@@ -45,17 +48,163 @@ export class ReferenceTable extends LookupTable {
+      const newRowsForResource = newRowsByResource.get(resource.id) ?? [];
+      const newHashes = new Set(newRowsForResource.map(hashRow));
+
+      const identical = existingHashes.size === newHashes.size && [...existingHashes].every((h) => newHashes.has(h));
+      if (!identical) {
+        resourcesToDelete.push(resource);
+        rowsToInsert.push(...newRowsForResource);
+      }
+    }
+
+    if (resourcesToDelete.length > 0) {
+      getLogger().info('Reference changes detected', {
+        resourceType,
+        unchangedCount: resources.length - resourcesToDelete.length,
+        changedCount: resourcesToDelete.length,
+        rowsToInsert: rowsToInsert.length,
+        sampleIds: resourcesToDelete.slice(0, 5),
+      });
+    }
+
+    if (resourcesToDelete.length > 0) {
+      await this.batchDeleteValuesForResources(client, resourcesToDelete);
+    }
+
+    if (rowsToInsert.length > 0) {
+      await this.batchInsertRows(client, resourceType, rowsToInsert);
+    }
+  }
+
+  /**
+   * Extracts values from all resources with batching to avoid blocking the event loop.
+   * @param result - The array to populate with extracted values.
+   * @param resources - The resources to extract values from.
+   */
+  private async extractAllValues<T extends Resource>(
+    result: ReferenceTableRow[] | Map<string, ReferenceTableRow[]>,
+    resources: WithId<T>[]
+  ): Promise<void> {
+    if (resources.length === 0) {
+      return;
+    }
+
+    const resourceType = resources[0].resourceType;
+
+    // Batch at the resource level to avoid tying up the event loop for too long
+    // with synchronous work without any async breaks between DB calls.
+    const resourceBatchSize = 200;
+    for (let i = 0; i < resources.length; i += resourceBatchSize) {
+      for (let j = i; j < i + resourceBatchSize && j < resources.length; j++) {
+        const resource = resources[j];
+        if (resource.resourceType !== resourceType) {
+          throw new Error(
+            `batchIndexResources must be called with resources of the same type: ${resource.resourceType} vs ${resourceType}`
+          );
+        }
+        try {
+          if (result instanceof Map) {
+            let rowArray = result.get(resource.id);
+            if (!rowArray) {
+              rowArray = [];
+              result.set(resource.id, rowArray);
+            }
+            this.extractValues(rowArray, resource);
+          } else {
+            this.extractValues(result, resource);
+          }
+        } catch (err) {
+          getLogger().error('Error extracting values for resource', {
+            resource: `${resourceType}/${resource.id}`,
+            err,
+          });
+          throw err;
+        }
+
+        // Yield to event loop between batches
+        if (i + resourceBatchSize < resources.length) {
+          await new Promise<void>((resolve) => {
+            setImmediate(() => resolve());
+          });
+        }
```

Comment:
This looks to me like it is inside the nested `j` loop and yielding for every record instead of each batch - should this be hoisted up one level into the `i` loop?